<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>FHIR Genomics Analysis for Clinical Decision Support – FHIR® for Research</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../workshops/data-analysis-cds/01-fhir-analysis.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dece3bd051e391dd7205a6b15f93dc59.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-ed04999718f06b735b1f8009dce43b94.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-84d8fd605c28f5564ab54bac3b8b1648.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark-9822177fbb414652ed72bbec35662ceb.min.css" rel="prefetch" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<!-- head.html -->
<!-- any content in this file is inserted into <head></head> -->

<script src="../../modules/mappings.js"></script>

<script>
// test if local storage is available and inititilize global variable store, code adapted from:
// https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
function initLocalStorage() {
  let storage;
  try {
    storage = window["localStorage"];
    const x = "__storage_test__";
    storage.setItem(x, x);
    storage.removeItem(x);
    return storage;
  } catch (e) {
    console.log("Error: local storage unavailable.");
    return (
      e instanceof DOMException &&
      // everything except Firefox
      (e.code === 22 ||
        // Firefox
        e.code === 1014 ||
        // test name field too, because code might not be present
        // everything except Firefox
        e.name === "QuotaExceededError" ||
        // Firefox
        e.name === "NS_ERROR_DOM_QUOTA_REACHED") &&
          // acknowledge QuotaExceededError only if there's something already stored
          storage &&
          storage.length !== 0
    );
  }
}

var store = initLocalStorage();
</script>

<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Yantramanav:wght@100;300;400;500;700;900&amp;display=swap" rel="stylesheet">

<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
<meta property="og:title" content="FHIR Genomics Analysis for Clinical Decision Support – FHIR® for Research">
<meta property="og:description" content="">
<meta property="og:site_name" content="FHIR® for Research">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../..">
    <span class="navbar-title">FHIR® for Research</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../sections/introduction"> 
<span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../sections/advanced-topics"> 
<span class="menu-text">Advanced Topics</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../sections/hands-on-practice"> 
<span class="menu-text">Hands-on Practice</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../sections/hosting-your-own-workshop"> 
<span class="menu-text">Hosting Your Own Workshop</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../contribute"> 
<span class="menu-text">Questions &amp; Contributions</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NIH-ODSS/fhir-for-research"> <i class="bi bi-github" role="img" aria-label="View the GitHub repository for this site">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../workshops/data-analysis-cds">FHIR Analysis, CDS and Genomics Workshop</a></li><li class="breadcrumb-item"><a href="../../workshops/data-analysis-cds/02-genomics-cds">Part 2: Genomics CDS</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">FHIR Analysis, CDS and Genomics Workshop</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../workshops/data-analysis-cds" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../workshops/data-analysis-cds/01-fhir-analysis" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Part 1: FHIR Data Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../workshops/data-analysis-cds/02-genomics-cds" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Part 2: Genomics CDS</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">SMART on FHIR</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../workshops/smart-on-fhir" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">FHIR Bulk Data</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../workshops/bulk-data" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#learning-paths" id="toc-learning-paths" class="nav-link" data-scroll-target="#learning-paths">Learning Paths</a></li>
  <li><a href="#basic-setup-and-connection-beginner-level" id="toc-basic-setup-and-connection-beginner-level" class="nav-link" data-scroll-target="#basic-setup-and-connection-beginner-level">Basic Setup and Connection (Beginner Level)</a></li>
  <li><a href="#analyzing-genetic-variants-intermediate-level" id="toc-analyzing-genetic-variants-intermediate-level" class="nav-link" data-scroll-target="#analyzing-genetic-variants-intermediate-level">Analyzing Genetic Variants (Intermediate Level)</a></li>
  <li><a href="#implementing-clinical-decision-support-advanced-level" id="toc-implementing-clinical-decision-support-advanced-level" class="nav-link" data-scroll-target="#implementing-clinical-decision-support-advanced-level">Implementing Clinical Decision Support (Advanced Level)</a>
  <ul class="collapse">
  <li><a href="#what-are-cds-hooks" id="toc-what-are-cds-hooks" class="nav-link" data-scroll-target="#what-are-cds-hooks">What Are CDS Hooks?</a></li>
  <li><a href="#implementing-a-cds-hooks-service-for-genomic-analysis" id="toc-implementing-a-cds-hooks-service-for-genomic-analysis" class="nav-link" data-scroll-target="#implementing-a-cds-hooks-service-for-genomic-analysis">Implementing a CDS Hooks Service for Genomic Analysis</a></li>
  <li><a href="#testing-our-cds-hooks-service" id="toc-testing-our-cds-hooks-service" class="nav-link" data-scroll-target="#testing-our-cds-hooks-service">Testing Our CDS Hooks Service</a></li>
  </ul></li>
  <li><a href="#summary-and-next-steps" id="toc-summary-and-next-steps" class="nav-link" data-scroll-target="#summary-and-next-steps">Summary and Next Steps</a></li>
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources">Additional Resources</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/nih-odss/fhir-for-research/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div><div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="02-genomics-cds.ipynb" download="02-genomics-cds.ipynb"><i class="bi bi-journal-code"></i>Jupyter</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../workshops/data-analysis-cds">FHIR Analysis, CDS and Genomics Workshop</a></li><li class="breadcrumb-item"><a href="../../workshops/data-analysis-cds/02-genomics-cds">Part 2: Genomics CDS</a></li></ol></nav><h1 class="title display-7">FHIR Genomics Analysis for Clinical Decision Support</h1></header>

<header id="title-block-header">
<!--
-->



</header>

<!-- Begin learning objectives block -->

<div class="callout-tip callout callout-style-default callout-captioned" id="learning-objectives">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Learning objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>
    Learn how to retrieve and analyze genomic data from FHIR servers using the Synthea Coherent Data Set
</li>
<li>
    Understand how to process genetic variants for clinical decision support
</li>
<li>
    Implement filtering and analysis of clinically significant variants
</li>
<li>
    Create CDS-ready outputs from FHIR genomic data
</li>
</ol>
</div>
</div>

<!-- End learning objectives block -->

<!-- Hack to add key points block at end of page-->
<!-- End hack to add key points block at end of page-->

<!-- Reading time logic -->
<script>
  document.addEventListener("DOMContentLoaded", function() {
  
    // Step 1: Count words in parent document
    let totalWords = 0;
    const mainContent = document.querySelector('#quarto-document-content');
    if (mainContent) {
      const parentWords = mainContent.innerText.trim().split(/\s+/).length;
      totalWords += parentWords;
    }
  
    // We'll update the badge once we've processed ALL iframes (or there are none)
    function updateReadingTimeBadge(wordCount) {
      const wpm = 200;
      const readingTime = Math.ceil(wordCount / wpm);
  
      let displayText;
      if (readingTime === 1) {
        displayText = 'Estimated reading time: 1 minute';
      } else {
        displayText = 'Estimated reading time: ' + readingTime + ' minutes';
      }
  
      // Create the badge
      const timeBadge = document.createElement('span');
      timeBadge.classList.add('reading-time-badge');
      timeBadge.textContent = displayText;

      const titleBlockHeader = document.querySelector('#title-block-header');
      if (titleBlockHeader) {
        titleBlockHeader.appendChild(timeBadge);
      }
    }
  
    // Step 2: Gather iframes
    const iframes = document.querySelectorAll('iframe');
    
    // If there are no iframes, just show the reading time for parent content
    if (iframes.length === 0) {
      updateReadingTimeBadge(totalWords);
      return;
    }
  
    // We only update after ALL iframes have loaded or errored
    let pending = iframes.length;
  
    // Helper to decrement pending count and update once done
    function iframeDone() {
      pending--;
      if (pending === 0) {
        // All iframes processed
        updateReadingTimeBadge(totalWords);
      }
    }
  
    // Step 3: Iterate through each iframe
    iframes.forEach(iframe => {
      // Listen for 'load' (fires whether same-origin or cross-origin)
      iframe.addEventListener('load', function() {
        try {
          // Attempt to read the iframe's text (works only if same-origin)
          const iframeDoc = iframe.contentWindow.document;
          const iframeText = (iframeDoc.body.innerText || "").trim();
          const iframeWords = iframeText.split(/\s+/).filter(Boolean).length;
  
          totalWords += iframeWords;
        } catch (err) {
          // Cross-origin or any other error
          console.warn("Could not read iframe content (possibly cross-origin):", err);
          // No addition to totalWords
        }
        // Mark this iframe as processed
        iframeDone();
      }, { once: true });
  
      // Listen for 'error' in case the iframe fails to load
      iframe.addEventListener('error', function() {
        console.warn("Iframe failed to load:", iframe.src);
        // Just move on (no addition to totalWords)
        iframeDone();
      }, { once: true });
    });
  
  });
  </script>



<p><a href="https://colab.research.google.com/github/NIH-ODSS/fhir-for-research/blob/main/workshops/data-analysis-cds/data/02-genomics-cds.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" class="img-fluid" alt="Open In Colab"></a> <a href="https://nbviewer.jupyter.org/github/NIH-ODSS/fhir-for-research/blob/main/workshops/data-analysis-cds/data/02-genomics-cds.ipynb"><img src="https://raw.githubusercontent.com/jupyter/design/refs/heads/main/logos/Badges/nbviewer_badge.svg" class="img-fluid" alt="View in NBViewer"></a> <!-- Force download instead of opening the notebook as a page--> <a href="./02-genomics-cds.ipynb" download=""><img src="https://img.shields.io/badge/download-notebook-F37626?logo=jupyter" alt="Download Notebook"></a> <a href="https://mybinder.org/v2/gh/NIH-ODSS/fhir-for-research/HEAD?urlpath=/doc/tree/workshops/data-analysis-cds/02-genomics-cds.ipynb"><img src="https://mybinder.org/badge_logo.svg" class="img-fluid" alt="Binder"></a></p>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>For the best learning experience, run this tutorial interactively, via one of the environment setup options. Use the above button depending on your chosen <a href="../../workshops/data-analysis-cds/index.html#environment-setup">setup option</a>.</p>
<p>This tutorial explores how to work with genomic data in FHIR for clinical decision support (CDS) applications. We’ll use the <a href="../../workshops/data-analysis-cds/index.html#synthetic-data">Synthea Coherent Data Set</a> to retrieve and analyze genomic reports and observations, ultimately implementing a CDS service that provides actionable recommendations based on a patient’s genetic profile.</p>
<p>In this tutorial, we’re using a FHIR server located at <strong>http://localhost:8080/fhir</strong>, but any FHIR server loaded with appropriate data can be used. For instructions on setting up your own test server, see <a href="../../modules/synthea-test-server">Standing up a FHIR Testing Server</a>.</p>
</section>
<section id="learning-paths" class="level2">
<h2 class="anchored" data-anchor-id="learning-paths">Learning Paths</h2>
<p>This tutorial offers three difficulty levels to accommodate different experience levels:</p>
<table class="caption-top table">
<caption>Difficulty Levels</caption>
<colgroup>
<col style="width: 20%">
<col style="width: 40%">
<col style="width: 40%">
</colgroup>
<thead>
<tr class="header">
<th>Level</th>
<th>Focus Areas</th>
<th>Recommended For</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="#basic-setup-and-connection-beginner-level">Beginner</a></td>
<td>FHIR server connection, Basic genomics report retrieval</td>
<td>Those new to FHIR Genomics</td>
</tr>
<tr class="even">
<td><a href="#analyzing-genetic-variants-intermediate-level">Intermediate</a></td>
<td>Variant analysis, Data transformation</td>
<td>Those familiar with genetic terminology and FHIR</td>
</tr>
<tr class="odd">
<td><a href="#implementing-clinical-decision-support-advanced-level">Advanced</a></td>
<td>CDS implementation, Clinical filtering</td>
<td>Experienced developers implementing genomic CDS systems</td>
</tr>
</tbody>
</table>
<p><strong>Prerequisites:</strong> Basic knowledge of Python, FHIR resources, and genomics terminology is recommended.</p>
</section>
<section id="basic-setup-and-connection-beginner-level" class="level2">
<h2 class="anchored" data-anchor-id="basic-setup-and-connection-beginner-level">Basic Setup and Connection (Beginner Level)</h2>
<p>In this section, you’ll learn how to:</p>
<ul>
<li>Connect to a FHIR server containing genomic data</li>
<li>Retrieve basic genetic diagnostic reports</li>
<li>Display genomic information in a structured format</li>
</ul>
<p>First, let’s set up our connection to the FHIR server:</p>
<div id="8f8ad910" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load dependency</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> requests, os</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>fhir_server <span class="op">=</span> os.environ.get(<span class="st">'FHIR_SERVER'</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Using FHIR server: </span><span class="sc">{</span>fhir_server<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if the server is running and connection is successful</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>response <span class="op">=</span> requests.get(<span class="ss">f"</span><span class="sc">{</span>fhir_server<span class="sc">}</span><span class="ss">/metadata"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Server status: </span><span class="sc">{</span>response<span class="sc">.</span>status_code<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Using FHIR server: http://localhost:8080/fhir
Server status: 200</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Understanding the FHIR Metadata Endpoint
</div>
</div>
<div class="callout-body-container callout-body">
<p>The metadata endpoint (<code>/metadata</code>) is a special FHIR endpoint that returns the server’s capability statement - a structured document that describes what the server can do. When we query this endpoint:</p>
<ul>
<li>We’re checking if the server is responsive (status code 200)</li>
<li>We’re verifying it’s a valid FHIR server</li>
<li>The response contains details about supported resources, operations, and search parameters</li>
</ul>
<p>This is a lightweight way to validate connectivity before attempting more complex queries.</p>
</div>
</div>
<p>If the server is responsive (code 200), proceed with the next code block.</p>
<div id="80294189" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import required libraries</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> json</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> requests</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> fhir_pyrate <span class="im">import</span> Pirate</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> display, HTML</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize FHIR-PYrate</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>search <span class="op">=</span> Pirate(</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    auth<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>    base_url<span class="op">=</span>fhir_server,</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    print_request_url<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s retrieve basic genetic diagnostic reports. The LOINC code <a href="https://loinc.org/55232-3/"><strong>55232-3</strong></a> refers to <em>Genetic analysis summary panel</em>, which is commonly used to identify genomic reports. To learn more about FHIRPath and mapping columns, refer to the intermediate level <a href="../../workshops/data-analysis-cds/01-fhir-analysis.html#selecting-specific-columns-intermediate-level">FHIR analysis exercise</a>.</p>
<div id="83a3f752" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fetch genomics reports</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Retrieving basic genetic diagnostic reports..."</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>reports_df <span class="op">=</span> search.steal_bundles_to_dataframe(</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    resource_type<span class="op">=</span><span class="st">"DiagnosticReport"</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    request_params<span class="op">=</span>{<span class="st">"code"</span>: <span class="st">"http://loinc.org|55232-3"</span>, <span class="st">"_count"</span>: <span class="dv">10</span>},</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    num_pages<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    fhir_paths<span class="op">=</span>[</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>        (<span class="st">"id"</span>, <span class="st">"id"</span>),</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>        (<span class="st">"patient"</span>, <span class="st">"subject.reference"</span>),</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>        (<span class="st">"status"</span>, <span class="st">"status"</span>),</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>        (<span class="st">"code_display"</span>, <span class="st">"code.coding[0].display"</span>),</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>        (<span class="st">"issued"</span>, <span class="st">"issued"</span>),</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the first few genomics reports</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"First few genomics reports:"</span>)</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>display(reports_df.head())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Retrieving basic genetic diagnostic reports...
http://localhost:8080/fhir/DiagnosticReport?_count=10&amp;code=http://loinc.org|55232-3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Query &amp; Build DF (DiagnosticReport):   0%|          | 0/1 [00:00&lt;?, ?it/s]Query &amp; Build DF (DiagnosticReport): 100%|██████████| 1/1 [00:00&lt;00:00, 213.65it/s]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>First few genomics reports:</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">id</th>
<th data-quarto-table-cell-role="th">patient</th>
<th data-quarto-table-cell-role="th">status</th>
<th data-quarto-table-cell-role="th">code_display</th>
<th data-quarto-table-cell-role="th">issued</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>53db09ba-0414-61e1-78d2-02b20a7092b8</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>final</td>
<td>Genetic analysis summary panel</td>
<td>2001-02-25T10:39:54.395-05:00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>c012aa9a-fe79-40bc-00ae-162c9906c6f8</td>
<td>Patient/39533e4a-f6f2-a144-ab37-6500460250dc</td>
<td>final</td>
<td>Genetic analysis summary panel</td>
<td>2017-02-21T03:33:40.769-05:00</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>27e51912-a805-26dc-b9c8-d88f131969a6</td>
<td>Patient/68c3ae0b-e298-62b7-5d3a-7936fd998fe0</td>
<td>final</td>
<td>Genetic analysis summary panel</td>
<td>1989-10-03T04:33:40.769-04:00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>a31d8d94-6acd-a5f0-618a-f1b38746f9e4</td>
<td>Patient/df860bc2-1943-237f-7445-ed960a1ef069</td>
<td>final</td>
<td>Genetic analysis summary panel</td>
<td>1992-02-10T08:04:51.390-05:00</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>1e5e18ae-bad4-3a45-b4e3-d2c7b7da66d9</td>
<td>Patient/7d9ba758-f0b8-3fc3-befa-f0e8e8fb6935</td>
<td>final</td>
<td>Genetic analysis summary panel</td>
<td>2007-11-23T16:23:59.928-05:00</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="genomics-concepts" class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Key Genomics Concepts in FHIR
</div>
</div>
<div class="callout-body-container callout-body">
<p>This tutorial introduces key FHIR genomics concepts from the <a href="http://hl7.org/fhir/uv/genomics-reporting/">HL7 FHIR Genomics Reporting IG</a>:</p>
<table class="caption-top table">
<caption>Key Genomics Concepts in FHIR</caption>
<colgroup>
<col style="width: 26%">
<col style="width: 34%">
<col style="width: 38%">
</colgroup>
<thead>
<tr class="header">
<th>Category</th>
<th>Description</th>
<th>FHIR Resource</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Genomic Report</td>
<td>Container for all genomic results</td>
<td>DiagnosticReport</td>
</tr>
<tr class="even">
<td>Genomic Findings</td>
<td>Variants, genotypes, and haplotypes</td>
<td>Observation profiles</td>
</tr>
<tr class="odd">
<td>Genomic Implications</td>
<td>Clinical impact of findings</td>
<td>Observation (linked via derivedFrom)</td>
</tr>
<tr class="even">
<td>Recommended Actions</td>
<td>Follow-up or treatment suggestions</td>
<td>Task</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Let’s examine what information we got back. The reports DataFrame contains basic information about genetic diagnostic reports, including:</p>
<ul>
<li>Patient references</li>
<li>Report status</li>
<li>Test types</li>
<li>Issue dates</li>
</ul>
</section>
<section id="analyzing-genetic-variants-intermediate-level" class="level2">
<h2 class="anchored" data-anchor-id="analyzing-genetic-variants-intermediate-level">Analyzing Genetic Variants (Intermediate Level)</h2>
<p>In this section, we’ll:</p>
<ul>
<li>Retrieve detailed genetic observations</li>
<li>Extract variant information</li>
<li>Process genetic data for clinical use</li>
</ul>
<p>Let’s retrieve diagnostic reports with their related observations in a single query:</p>
<div id="38ee5eb2" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Retrieve reports with related observations</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Retrieving genetic reports with observations..."</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>results <span class="op">=</span> search.steal_bundles_to_dataframe(</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    resource_type<span class="op">=</span><span class="st">"DiagnosticReport"</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    request_params<span class="op">=</span>{</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"code"</span>: <span class="st">"http://loinc.org|55232-3"</span>,  <span class="co"># LOINC code for "Genetic analysis summary panel"</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"_include"</span>: <span class="st">"DiagnosticReport:result"</span>,  <span class="co"># Include the Observation resources</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"_count"</span>: <span class="dv">10</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>    num_pages<span class="op">=</span><span class="dv">1</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the different resource types from the results</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>reports_df <span class="op">=</span> results.get(<span class="st">"DiagnosticReport"</span>, pd.DataFrame())</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>observations_df <span class="op">=</span> results.get(<span class="st">"Observation"</span>, pd.DataFrame())</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Display basic report information</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">Retrieved genetic analysis reports:"</span>)</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>display(reports_df[[<span class="st">"id"</span>, <span class="st">"subject_reference"</span>, <span class="st">"status"</span>, <span class="st">"issued"</span>]].head())</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Display observation information</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">Retrieved genetic observations:"</span>)</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Found </span><span class="sc">{</span><span class="bu">len</span>(observations_df)<span class="sc">}</span><span class="ss"> observations"</span>)</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>display(observations_df.head(<span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Retrieving genetic reports with observations...
http://localhost:8080/fhir/DiagnosticReport?_count=10&amp;_include=DiagnosticReport:result&amp;code=http://loinc.org|55232-3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Query &amp; Build DF (DiagnosticReport):   0%|          | 0/1 [00:00&lt;?, ?it/s]Query &amp; Build DF (DiagnosticReport): 100%|██████████| 1/1 [00:00&lt;00:00, 152.00it/s]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Retrieved genetic analysis reports:</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">id</th>
<th data-quarto-table-cell-role="th">subject_reference</th>
<th data-quarto-table-cell-role="th">status</th>
<th data-quarto-table-cell-role="th">issued</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>53db09ba-0414-61e1-78d2-02b20a7092b8</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>final</td>
<td>2001-02-25T10:39:54.395-05:00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>c012aa9a-fe79-40bc-00ae-162c9906c6f8</td>
<td>Patient/39533e4a-f6f2-a144-ab37-6500460250dc</td>
<td>final</td>
<td>2017-02-21T03:33:40.769-05:00</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>27e51912-a805-26dc-b9c8-d88f131969a6</td>
<td>Patient/68c3ae0b-e298-62b7-5d3a-7936fd998fe0</td>
<td>final</td>
<td>1989-10-03T04:33:40.769-04:00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>a31d8d94-6acd-a5f0-618a-f1b38746f9e4</td>
<td>Patient/df860bc2-1943-237f-7445-ed960a1ef069</td>
<td>final</td>
<td>1992-02-10T08:04:51.390-05:00</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>1e5e18ae-bad4-3a45-b4e3-d2c7b7da66d9</td>
<td>Patient/7d9ba758-f0b8-3fc3-befa-f0e8e8fb6935</td>
<td>final</td>
<td>2007-11-23T16:23:59.928-05:00</td>
</tr>
</tbody>
</table>

</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Retrieved genetic observations:
Found 159 observations</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">resourceType</th>
<th data-quarto-table-cell-role="th">id</th>
<th data-quarto-table-cell-role="th">meta_versionId</th>
<th data-quarto-table-cell-role="th">meta_lastUpdated</th>
<th data-quarto-table-cell-role="th">meta_source</th>
<th data-quarto-table-cell-role="th">status</th>
<th data-quarto-table-cell-role="th">category_0_coding_0_system</th>
<th data-quarto-table-cell-role="th">code_coding_0_system</th>
<th data-quarto-table-cell-role="th">code_coding_0_code</th>
<th data-quarto-table-cell-role="th">code_coding_0_display</th>
<th data-quarto-table-cell-role="th">code_text</th>
<th data-quarto-table-cell-role="th">subject_reference</th>
<th data-quarto-table-cell-role="th">encounter_reference</th>
<th data-quarto-table-cell-role="th">effectiveDateTime</th>
<th data-quarto-table-cell-role="th">issued</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>Observation</td>
<td>526aeec5-b820-826e-2ea5-133be88c496b</td>
<td>1</td>
<td>2025-05-22T21:25:11.670+00:00</td>
<td>#cRLjd7OxPBpSom2g</td>
<td>final</td>
<td>http://terminology.hl7.org/CodeSystem/observat...</td>
<td>http://loinc.org</td>
<td>69548-6</td>
<td>The LPA gene exhibits a variation of 'Uncertai...</td>
<td>The LPA gene exhibits a variation of 'Uncertai...</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>Encounter/8c2ce4f9-3f70-c405-09d1-48f5992f22e9</td>
<td>2001-02-25T10:39:54-05:00</td>
<td>2001-02-25T10:39:54.395-05:00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>Observation</td>
<td>d226080a-003b-973d-97f0-ba94ff92a7b8</td>
<td>1</td>
<td>2025-05-22T21:25:11.670+00:00</td>
<td>#cRLjd7OxPBpSom2g</td>
<td>final</td>
<td>http://terminology.hl7.org/CodeSystem/observat...</td>
<td>http://loinc.org</td>
<td>69548-6</td>
<td>The LPA gene exhibits a variation of 'Uncertai...</td>
<td>The LPA gene exhibits a variation of 'Uncertai...</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>Encounter/8c2ce4f9-3f70-c405-09d1-48f5992f22e9</td>
<td>2001-02-25T10:39:54-05:00</td>
<td>2001-02-25T10:39:54.395-05:00</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>Observation</td>
<td>18c160fa-1524-7d7d-0bbb-c9cb7b6b19c0</td>
<td>1</td>
<td>2025-05-22T21:25:11.670+00:00</td>
<td>#cRLjd7OxPBpSom2g</td>
<td>final</td>
<td>http://terminology.hl7.org/CodeSystem/observat...</td>
<td>http://loinc.org</td>
<td>69548-6</td>
<td>The LOC729065 gene exhibits a variation of 'Un...</td>
<td>The LOC729065 gene exhibits a variation of 'Un...</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>Encounter/8c2ce4f9-3f70-c405-09d1-48f5992f22e9</td>
<td>2001-02-25T10:39:54-05:00</td>
<td>2001-02-25T10:39:54.395-05:00</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="fhir-search-params" class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
FHIR Search Parameters for Genomic Data
</div>
</div>
<div class="callout-body-container callout-body">
<p>When working with the Synthea Coherent Data Set:</p>
<ul>
<li>Use LOINC code <strong>55232-3</strong> for genetic reports</li>
<li>Use <code>_include=DiagnosticReport:result</code> to retrieve related observations</li>
<li>The dataset follows <a href="https://build.fhir.org/ig/HL7/US-Core/StructureDefinition-us-core-diagnosticreport-lab.html">US Core profiles for lab reports</a> rather than <a href="https://build.fhir.org/ig/HL7/genomics-reporting/StructureDefinition-genomic-report.html">specialized genomic profiles</a>.</li>
</ul>
</div>
</div>
<p>Now, let’s extract genetic variant information from the observations:</p>
<div id="3cdab34a" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract genetic variant information from observations</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">Extracting genetic variant information..."</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a list to store variant information</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>variants <span class="op">=</span> []</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Process each observation to extract key information</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> _, row <span class="kw">in</span> observations_df.iterrows():</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Check if text content is available</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">'code_text'</span> <span class="kw">in</span> row <span class="kw">and</span> <span class="kw">not</span> pd.isna(row[<span class="st">'code_text'</span>]):</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>        display_text <span class="op">=</span> <span class="bu">str</span>(row[<span class="st">'code_text'</span>])</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Extract genetic information using regular expressions</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>        gene_match <span class="op">=</span> re.search(<span class="vs">r'The (\S+) gene'</span>, display_text)</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>        significance_match <span class="op">=</span> re.search(<span class="vs">r"variation of '([^']+)'"</span>, display_text)</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>        variant_match <span class="op">=</span> re.search(<span class="vs">r'index (\S+) is'</span>, display_text)</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>        risks_match <span class="op">=</span> re.search(<span class="vs">r'risk of: (.+)\.'</span>, display_text)</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Create a record with the extracted information</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>        variant_info <span class="op">=</span> {</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>            <span class="st">'patient'</span>: row[<span class="st">'subject_reference'</span>],</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>            <span class="st">'gene'</span>: gene_match.group(<span class="dv">1</span>) <span class="cf">if</span> gene_match <span class="cf">else</span> <span class="st">'Unknown'</span>,</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>            <span class="st">'clinical_significance'</span>: significance_match.group(<span class="dv">1</span>) <span class="cf">if</span> significance_match <span class="cf">else</span> <span class="st">'Unknown'</span>,</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>            <span class="st">'variant_id'</span>: variant_match.group(<span class="dv">1</span>) <span class="cf">if</span> variant_match <span class="cf">else</span> <span class="st">'Unknown'</span>,</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>            <span class="st">'associated_risks'</span>: risks_match.group(<span class="dv">1</span>) <span class="cf">if</span> risks_match <span class="cf">else</span> <span class="st">'Unknown'</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>        variants.append(variant_info)</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to DataFrame for easier analysis</span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>variants_df <span class="op">=</span> pd.DataFrame(variants)</span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the extracted variants</span></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">Extracted genetic variants:"</span>)</span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>display(variants_df.head(<span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Extracting genetic variant information...

Extracted genetic variants:</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">patient</th>
<th data-quarto-table-cell-role="th">gene</th>
<th data-quarto-table-cell-role="th">clinical_significance</th>
<th data-quarto-table-cell-role="th">variant_id</th>
<th data-quarto-table-cell-role="th">associated_risks</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>LPA</td>
<td>Uncertain</td>
<td>rs10033464_FS1</td>
<td>Ischemic Stroke</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>LPA</td>
<td>Uncertain</td>
<td>rs10033464_FS2</td>
<td>Ischemic Stroke</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>LOC729065</td>
<td>Uncertain</td>
<td>rs2200733_FS2</td>
<td>Ischemic Stroke and Stroke</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>ADD1</td>
<td>Risk Factor</td>
<td>rs4961_FS1</td>
<td>Cardiovascular Issues, High LDL, Hemorrhagic S...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>ADD1</td>
<td>Risk Factor</td>
<td>rs4961_FS2</td>
<td>Cardiovascular Issues, High LDL, Hemorrhagic S...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="implementing-clinical-decision-support-advanced-level" class="level2">
<h2 class="anchored" data-anchor-id="implementing-clinical-decision-support-advanced-level">Implementing Clinical Decision Support (Advanced Level)</h2>
<p>In this section, we’ll:</p>
<ul>
<li>Filter for clinically significant variants</li>
<li>Generate clinical recommendations</li>
<li>Implement a CDS service to deliver these recommendations</li>
</ul>
<p>First, let’s filter for variants with significant clinical impact:</p>
<div id="f9b1be2d" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for variants with significant clinical impact</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Filtering for clinically significant variants..."</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>pathogenic_variants <span class="op">=</span> variants_df[</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    variants_df[<span class="st">'clinical_significance'</span>].<span class="bu">str</span>.contains(<span class="st">'Pathogenic|Risk Factor'</span>, na<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Found </span><span class="sc">{</span><span class="bu">len</span>(pathogenic_variants)<span class="sc">}</span><span class="ss"> variants with clinical significance"</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>display(pathogenic_variants.head())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtering for clinically significant variants...
Found 25 variants with clinical significance</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">patient</th>
<th data-quarto-table-cell-role="th">gene</th>
<th data-quarto-table-cell-role="th">clinical_significance</th>
<th data-quarto-table-cell-role="th">variant_id</th>
<th data-quarto-table-cell-role="th">associated_risks</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>ADD1</td>
<td>Risk Factor</td>
<td>rs4961_FS1</td>
<td>Cardiovascular Issues, High LDL, Hemorrhagic S...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>ADD1</td>
<td>Risk Factor</td>
<td>rs4961_FS2</td>
<td>Cardiovascular Issues, High LDL, Hemorrhagic S...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>PON1</td>
<td>Risk Factor</td>
<td>rs662_FS1</td>
<td>Cardiovascular Issues, Mitral valve issues, Ds...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">10</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>PON1</td>
<td>Risk Factor</td>
<td>rs662_FS2</td>
<td>Cardiovascular Issues, Mitral valve issues, Ds...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">11</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>PON1</td>
<td>Risk Factor</td>
<td>rs854560_FS2</td>
<td>Cardiovascular Issues, Insulin Resistance, Low...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Now that we’ve analyzed genetic variants and generated recommendations, let’s explore how to deploy these as a clinical decision support service using CDS Hooks.</p>
<section id="what-are-cds-hooks" class="level3">
<h3 class="anchored" data-anchor-id="what-are-cds-hooks">What Are CDS Hooks?</h3>
<p>CDS Hooks is a standard for integrating clinical decision support into electronic health record (EHR) systems at the point of care. It allows our genomic analysis recommendations to be delivered to clinicians at the right time and in the right context.</p>
<p>For a comprehensive introduction to CDS Hooks, please refer to <a href="../../modules/new-cds-hooks-intro">CDS Hooks Introduction</a>, which covers the fundamental concepts, workflow, and implementation details.</p>
<div id="cds-hooks-overview" class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
CDS Hooks at a Glance
</div>
</div>
<div class="callout-body-container callout-body">
<p>CDS Hooks provides a framework for:</p>
<ol type="1">
<li><strong>Event-Driven Integration</strong>: Hooks trigger at specific points in clinical workflow</li>
<li><strong>Standard API</strong>: Consistent way for EHRs to communicate with CDS services</li>
<li><strong>Card-Based Interface</strong>: Recommendations delivered as “cards” within EHR UI</li>
<li><strong>Context-Aware</strong>: Services receive patient context to provide relevant advice</li>
</ol>
<p><strong>Common Hook Points:</strong></p>
<ul>
<li><em>patient-view</em>: When a patient’s record is opened</li>
<li><em>order-select</em>: When a medication or procedure is selected</li>
<li><em>order-sign</em>: When orders are about to be signed</li>
</ul>
<p><strong>The CDS Hooks Workflow</strong></p>
<p>When integrated into a real clinical environment, the CDS Hooks workflow would look like this:</p>
<ol type="1">
<li><strong>Trigger</strong>: A clinician opens a patient’s record, triggering the <code>patient-view</code> hook</li>
<li><strong>Request</strong>: The EHR sends a request to a CDS service with the patient context</li>
<li><strong>Analysis</strong>: CDS service queries the FHIR server for relevant genomic data and analyzes variants</li>
<li><strong>Response</strong>: CDS service returns CDS Hooks cards with clinical recommendations</li>
<li><strong>Display</strong>: The EHR displays these recommendations to the clinician within their workflow</li>
</ol>
</div>
</div>
</section>
<section id="implementing-a-cds-hooks-service-for-genomic-analysis" class="level3">
<h3 class="anchored" data-anchor-id="implementing-a-cds-hooks-service-for-genomic-analysis">Implementing a CDS Hooks Service for Genomic Analysis</h3>
<p>Now, let’s create a CDS Hooks service to serve our genomic recommendations. This service will:</p>
<ol type="1">
<li>Analyze a patient’s genetic data in real-time</li>
<li>Generate appropriate clinical recommendations</li>
<li>Return these recommendations as CDS Hooks cards</li>
</ol>
<p>The service exposes two endpoints:</p>
<ul>
<li>A discovery endpoint that tells EHRs what services are available</li>
<li>A service endpoint that processes requests and returns recommendations</li>
</ul>
<p>We will use <a href="https://flask.palletsprojects.com/en/stable/">Flask</a>, a lightweight application framework for writing web applications and services using Python.</p>
<div id="e382a1ac" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import necessary libraries</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> flask <span class="im">import</span> Flask, jsonify, request</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> threading</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> time</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Choose a port for the Flask server</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>flask_port <span class="op">=</span> <span class="dv">3299</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>flask_server_thread <span class="op">=</span> <span class="va">None</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a Flask application</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>app <span class="op">=</span> Flask(<span class="va">__name__</span>)</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="co"># CDS Hooks discovery endpoint - this tells EHRs what services we provide</span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="at">@app.route</span>(<span class="st">'/cds-services'</span>, methods<span class="op">=</span>[<span class="st">'GET'</span>])</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> discovery():</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> jsonify({</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>        <span class="st">'services'</span>: [</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>            {</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>                <span class="st">'id'</span>: <span class="st">'genomics-advisor'</span>,</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>                <span class="st">'hook'</span>: <span class="st">'patient-view'</span>,</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>                <span class="st">'title'</span>: <span class="st">'Genomics Clinical Advisor'</span>,</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>                <span class="st">'description'</span>: <span class="st">'Provides recommendations based on genetic findings'</span>,</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Note: In a production environment, prefetch could be used for optimization</span></span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>                <span class="co"># but is omitted here for simplicity</span></span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>        ]</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a><span class="co"># CDS Hooks service endpoint - this returns recommendations for a specific patient</span></span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a><span class="at">@app.route</span>(<span class="st">'/cds-services/genomics-advisor'</span>, methods<span class="op">=</span>[<span class="st">'POST'</span>])</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> genomics_advisor():</span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get data from the request</span></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a>    request_data <span class="op">=</span> request.json</span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a>    context <span class="op">=</span> request_data.get(<span class="st">'context'</span>, {})</span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a>    patient_id <span class="op">=</span> context.get(<span class="st">'patientId'</span>)</span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Processing request for patient: </span><span class="sc">{</span>patient_id<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Query for this specific patient's genetic reports</span></span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a>        diagnostic_reports <span class="op">=</span> search.steal_bundles_to_dataframe(</span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a>            resource_type<span class="op">=</span><span class="st">"DiagnosticReport"</span>,</span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a>            request_params<span class="op">=</span>{</span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a>                <span class="st">"code"</span>: <span class="st">"http://loinc.org|55232-3"</span>,  <span class="co"># Genetic analysis summary panel</span></span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a>                <span class="st">"subject"</span>: patient_id,</span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a>                <span class="st">"_count"</span>: <span class="dv">100</span></span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a>            num_pages<span class="op">=</span><span class="dv">1</span></span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Check if we got results</span></span>
<span id="cb19-54"><a href="#cb19-54" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">isinstance</span>(diagnostic_reports, <span class="bu">dict</span>) <span class="kw">and</span> <span class="st">"DiagnosticReport"</span> <span class="kw">in</span> diagnostic_reports:</span>
<span id="cb19-55"><a href="#cb19-55" aria-hidden="true" tabindex="-1"></a>            reports_df <span class="op">=</span> diagnostic_reports[<span class="st">"DiagnosticReport"</span>]</span>
<span id="cb19-56"><a href="#cb19-56" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb19-57"><a href="#cb19-57" aria-hidden="true" tabindex="-1"></a>            <span class="co"># The data might be directly in the DataFrame</span></span>
<span id="cb19-58"><a href="#cb19-58" aria-hidden="true" tabindex="-1"></a>            reports_df <span class="op">=</span> diagnostic_reports</span>
<span id="cb19-59"><a href="#cb19-59" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb19-60"><a href="#cb19-60" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> reports_df.empty:</span>
<span id="cb19-61"><a href="#cb19-61" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="st">"No genetic reports found for this patient"</span>)</span>
<span id="cb19-62"><a href="#cb19-62" aria-hidden="true" tabindex="-1"></a>            pathogenic_variants <span class="op">=</span> pd.DataFrame()</span>
<span id="cb19-63"><a href="#cb19-63" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb19-64"><a href="#cb19-64" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f"Found </span><span class="sc">{</span><span class="bu">len</span>(reports_df)<span class="sc">}</span><span class="ss"> genetic reports for patient </span><span class="sc">{</span>patient_id<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb19-65"><a href="#cb19-65" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb19-66"><a href="#cb19-66" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Extract genetic information from flattened result columns</span></span>
<span id="cb19-67"><a href="#cb19-67" aria-hidden="true" tabindex="-1"></a>            variants <span class="op">=</span> []</span>
<span id="cb19-68"><a href="#cb19-68" aria-hidden="true" tabindex="-1"></a>            <span class="im">import</span> re</span>
<span id="cb19-69"><a href="#cb19-69" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb19-70"><a href="#cb19-70" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Identify all result display columns</span></span>
<span id="cb19-71"><a href="#cb19-71" aria-hidden="true" tabindex="-1"></a>            display_columns <span class="op">=</span> [col <span class="cf">for</span> col <span class="kw">in</span> reports_df.columns <span class="cf">if</span> <span class="st">'result'</span> <span class="kw">in</span> col <span class="kw">and</span> <span class="st">'display'</span> <span class="kw">in</span> col]</span>
<span id="cb19-72"><a href="#cb19-72" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f"Found </span><span class="sc">{</span><span class="bu">len</span>(display_columns)<span class="sc">}</span><span class="ss"> result display columns"</span>)</span>
<span id="cb19-73"><a href="#cb19-73" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb19-74"><a href="#cb19-74" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Process each result display column</span></span>
<span id="cb19-75"><a href="#cb19-75" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> col <span class="kw">in</span> display_columns:</span>
<span id="cb19-76"><a href="#cb19-76" aria-hidden="true" tabindex="-1"></a>                <span class="cf">for</span> _, row <span class="kw">in</span> reports_df.iterrows():</span>
<span id="cb19-77"><a href="#cb19-77" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> pd.notna(row[col]):</span>
<span id="cb19-78"><a href="#cb19-78" aria-hidden="true" tabindex="-1"></a>                        display_text <span class="op">=</span> <span class="bu">str</span>(row[col])</span>
<span id="cb19-79"><a href="#cb19-79" aria-hidden="true" tabindex="-1"></a>                        </span>
<span id="cb19-80"><a href="#cb19-80" aria-hidden="true" tabindex="-1"></a>                        <span class="co"># Extract information using regex</span></span>
<span id="cb19-81"><a href="#cb19-81" aria-hidden="true" tabindex="-1"></a>                        gene_match <span class="op">=</span> re.search(<span class="vs">r'The (\S+) gene'</span>, display_text)</span>
<span id="cb19-82"><a href="#cb19-82" aria-hidden="true" tabindex="-1"></a>                        significance_match <span class="op">=</span> re.search(<span class="vs">r"variation of '([^']+)'"</span>, display_text)</span>
<span id="cb19-83"><a href="#cb19-83" aria-hidden="true" tabindex="-1"></a>                        variant_match <span class="op">=</span> re.search(<span class="vs">r'index (\S+) is'</span>, display_text)</span>
<span id="cb19-84"><a href="#cb19-84" aria-hidden="true" tabindex="-1"></a>                        risks_match <span class="op">=</span> re.search(<span class="vs">r'risk of: (.+)\.'</span>, display_text)</span>
<span id="cb19-85"><a href="#cb19-85" aria-hidden="true" tabindex="-1"></a>                        </span>
<span id="cb19-86"><a href="#cb19-86" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">if</span> gene_match <span class="kw">or</span> significance_match:</span>
<span id="cb19-87"><a href="#cb19-87" aria-hidden="true" tabindex="-1"></a>                            variant_info <span class="op">=</span> {</span>
<span id="cb19-88"><a href="#cb19-88" aria-hidden="true" tabindex="-1"></a>                                <span class="st">'gene'</span>: gene_match.group(<span class="dv">1</span>) <span class="cf">if</span> gene_match <span class="cf">else</span> <span class="st">'Unknown'</span>,</span>
<span id="cb19-89"><a href="#cb19-89" aria-hidden="true" tabindex="-1"></a>                                <span class="st">'clinical_significance'</span>: significance_match.group(<span class="dv">1</span>) <span class="cf">if</span> significance_match <span class="cf">else</span> <span class="st">'Unknown'</span>,</span>
<span id="cb19-90"><a href="#cb19-90" aria-hidden="true" tabindex="-1"></a>                                <span class="st">'variant_id'</span>: variant_match.group(<span class="dv">1</span>) <span class="cf">if</span> variant_match <span class="cf">else</span> <span class="st">'Unknown'</span>,</span>
<span id="cb19-91"><a href="#cb19-91" aria-hidden="true" tabindex="-1"></a>                                <span class="st">'associated_risks'</span>: risks_match.group(<span class="dv">1</span>) <span class="cf">if</span> risks_match <span class="cf">else</span> <span class="st">'Unknown'</span></span>
<span id="cb19-92"><a href="#cb19-92" aria-hidden="true" tabindex="-1"></a>                            }</span>
<span id="cb19-93"><a href="#cb19-93" aria-hidden="true" tabindex="-1"></a>                            </span>
<span id="cb19-94"><a href="#cb19-94" aria-hidden="true" tabindex="-1"></a>                            variants.append(variant_info)</span>
<span id="cb19-95"><a href="#cb19-95" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb19-96"><a href="#cb19-96" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Convert to DataFrame and filter for pathogenic variants</span></span>
<span id="cb19-97"><a href="#cb19-97" aria-hidden="true" tabindex="-1"></a>            variants_df <span class="op">=</span> pd.DataFrame(variants)</span>
<span id="cb19-98"><a href="#cb19-98" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f"Extracted </span><span class="sc">{</span><span class="bu">len</span>(variants_df)<span class="sc">}</span><span class="ss"> variants"</span>)</span>
<span id="cb19-99"><a href="#cb19-99" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb19-100"><a href="#cb19-100" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="kw">not</span> variants_df.empty:</span>
<span id="cb19-101"><a href="#cb19-101" aria-hidden="true" tabindex="-1"></a>                pathogenic_variants <span class="op">=</span> variants_df[</span>
<span id="cb19-102"><a href="#cb19-102" aria-hidden="true" tabindex="-1"></a>                    variants_df[<span class="st">'clinical_significance'</span>].<span class="bu">str</span>.contains(<span class="st">'Pathogenic|Risk Factor'</span>, case<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb19-103"><a href="#cb19-103" aria-hidden="true" tabindex="-1"></a>                ]</span>
<span id="cb19-104"><a href="#cb19-104" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="ss">f"Found </span><span class="sc">{</span><span class="bu">len</span>(pathogenic_variants)<span class="sc">}</span><span class="ss"> pathogenic variants"</span>)</span>
<span id="cb19-105"><a href="#cb19-105" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb19-106"><a href="#cb19-106" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="st">"No variants could be extracted"</span>)</span>
<span id="cb19-107"><a href="#cb19-107" aria-hidden="true" tabindex="-1"></a>                pathogenic_variants <span class="op">=</span> pd.DataFrame()</span>
<span id="cb19-108"><a href="#cb19-108" aria-hidden="true" tabindex="-1"></a>                </span>
<span id="cb19-109"><a href="#cb19-109" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span> <span class="pp">Exception</span> <span class="im">as</span> e:</span>
<span id="cb19-110"><a href="#cb19-110" aria-hidden="true" tabindex="-1"></a>        <span class="im">import</span> traceback</span>
<span id="cb19-111"><a href="#cb19-111" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Error querying FHIR server: </span><span class="sc">{</span>e<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb19-112"><a href="#cb19-112" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"Traceback:"</span>)</span>
<span id="cb19-113"><a href="#cb19-113" aria-hidden="true" tabindex="-1"></a>        traceback.print_exc()</span>
<span id="cb19-114"><a href="#cb19-114" aria-hidden="true" tabindex="-1"></a>        pathogenic_variants <span class="op">=</span> pd.DataFrame()</span>
<span id="cb19-115"><a href="#cb19-115" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-116"><a href="#cb19-116" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Generate recommendations based on pathogenic variants</span></span>
<span id="cb19-117"><a href="#cb19-117" aria-hidden="true" tabindex="-1"></a>    recommendations <span class="op">=</span> []</span>
<span id="cb19-118"><a href="#cb19-118" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-119"><a href="#cb19-119" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> pathogenic_variants.empty:</span>
<span id="cb19-120"><a href="#cb19-120" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Check for specific gene variants</span></span>
<span id="cb19-121"><a href="#cb19-121" aria-hidden="true" tabindex="-1"></a>        genes <span class="op">=</span> pathogenic_variants[<span class="st">'gene'</span>].unique()</span>
<span id="cb19-122"><a href="#cb19-122" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Found genes for recommendations: </span><span class="sc">{</span>genes<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb19-123"><a href="#cb19-123" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb19-124"><a href="#cb19-124" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Known genes with clinical recommendations</span></span>
<span id="cb19-125"><a href="#cb19-125" aria-hidden="true" tabindex="-1"></a>        gene_recommendations <span class="op">=</span> {</span>
<span id="cb19-126"><a href="#cb19-126" aria-hidden="true" tabindex="-1"></a>            <span class="st">'APOE'</span>: {</span>
<span id="cb19-127"><a href="#cb19-127" aria-hidden="true" tabindex="-1"></a>                <span class="st">'title'</span>: <span class="st">'APOE Pathogenic Variant Detected'</span>,</span>
<span id="cb19-128"><a href="#cb19-128" aria-hidden="true" tabindex="-1"></a>                <span class="st">'detail'</span>: <span class="st">'Consider lipid panel and cardiovascular risk assessment'</span>,</span>
<span id="cb19-129"><a href="#cb19-129" aria-hidden="true" tabindex="-1"></a>                <span class="st">'source'</span>: <span class="st">'Clinical Practice Guidelines'</span>,</span>
<span id="cb19-130"><a href="#cb19-130" aria-hidden="true" tabindex="-1"></a>                <span class="st">'urgency'</span>: <span class="st">'high'</span></span>
<span id="cb19-131"><a href="#cb19-131" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb19-132"><a href="#cb19-132" aria-hidden="true" tabindex="-1"></a>            <span class="st">'BRCA1'</span>: {</span>
<span id="cb19-133"><a href="#cb19-133" aria-hidden="true" tabindex="-1"></a>                <span class="st">'title'</span>: <span class="st">'BRCA1 Pathogenic Variant Detected'</span>,</span>
<span id="cb19-134"><a href="#cb19-134" aria-hidden="true" tabindex="-1"></a>                <span class="st">'detail'</span>: <span class="st">'Consider cancer risk assessment and screening'</span>,</span>
<span id="cb19-135"><a href="#cb19-135" aria-hidden="true" tabindex="-1"></a>                <span class="st">'source'</span>: <span class="st">'NCCN Guidelines'</span>,</span>
<span id="cb19-136"><a href="#cb19-136" aria-hidden="true" tabindex="-1"></a>                <span class="st">'urgency'</span>: <span class="st">'high'</span></span>
<span id="cb19-137"><a href="#cb19-137" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb19-138"><a href="#cb19-138" aria-hidden="true" tabindex="-1"></a>            <span class="st">'BRCA2'</span>: {</span>
<span id="cb19-139"><a href="#cb19-139" aria-hidden="true" tabindex="-1"></a>                <span class="st">'title'</span>: <span class="st">'BRCA2 Pathogenic Variant Detected'</span>,</span>
<span id="cb19-140"><a href="#cb19-140" aria-hidden="true" tabindex="-1"></a>                <span class="st">'detail'</span>: <span class="st">'Consider cancer risk assessment and screening'</span>,</span>
<span id="cb19-141"><a href="#cb19-141" aria-hidden="true" tabindex="-1"></a>                <span class="st">'source'</span>: <span class="st">'NCCN Guidelines'</span>,</span>
<span id="cb19-142"><a href="#cb19-142" aria-hidden="true" tabindex="-1"></a>                <span class="st">'urgency'</span>: <span class="st">'high'</span></span>
<span id="cb19-143"><a href="#cb19-143" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb19-144"><a href="#cb19-144" aria-hidden="true" tabindex="-1"></a>            <span class="st">'PON1'</span>: {</span>
<span id="cb19-145"><a href="#cb19-145" aria-hidden="true" tabindex="-1"></a>                <span class="st">'title'</span>: <span class="st">'PON1 Risk Variant Detected'</span>,</span>
<span id="cb19-146"><a href="#cb19-146" aria-hidden="true" tabindex="-1"></a>                <span class="st">'detail'</span>: <span class="st">'Consider cardiovascular risk assessment and lipid-lowering therapy'</span>,</span>
<span id="cb19-147"><a href="#cb19-147" aria-hidden="true" tabindex="-1"></a>                <span class="st">'source'</span>: <span class="st">'Cardiovascular Guidelines'</span>,</span>
<span id="cb19-148"><a href="#cb19-148" aria-hidden="true" tabindex="-1"></a>                <span class="st">'urgency'</span>: <span class="st">'medium'</span></span>
<span id="cb19-149"><a href="#cb19-149" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb19-150"><a href="#cb19-150" aria-hidden="true" tabindex="-1"></a>            <span class="st">'ADRB3'</span>: {</span>
<span id="cb19-151"><a href="#cb19-151" aria-hidden="true" tabindex="-1"></a>                <span class="st">'title'</span>: <span class="st">'ADRB3 Risk Variant Detected'</span>,</span>
<span id="cb19-152"><a href="#cb19-152" aria-hidden="true" tabindex="-1"></a>                <span class="st">'detail'</span>: <span class="st">'Associated with metabolic disorders and cardiovascular risk'</span>,</span>
<span id="cb19-153"><a href="#cb19-153" aria-hidden="true" tabindex="-1"></a>                <span class="st">'source'</span>: <span class="st">'Metabolic Risk Guidelines'</span>,</span>
<span id="cb19-154"><a href="#cb19-154" aria-hidden="true" tabindex="-1"></a>                <span class="st">'urgency'</span>: <span class="st">'medium'</span></span>
<span id="cb19-155"><a href="#cb19-155" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb19-156"><a href="#cb19-156" aria-hidden="true" tabindex="-1"></a>            <span class="st">'CCL2'</span>: {</span>
<span id="cb19-157"><a href="#cb19-157" aria-hidden="true" tabindex="-1"></a>                <span class="st">'title'</span>: <span class="st">'CCL2 Pathogenic Variant Detected'</span>,</span>
<span id="cb19-158"><a href="#cb19-158" aria-hidden="true" tabindex="-1"></a>                <span class="st">'detail'</span>: <span class="st">'Associated with inflammatory processes and stroke risk'</span>,</span>
<span id="cb19-159"><a href="#cb19-159" aria-hidden="true" tabindex="-1"></a>                <span class="st">'source'</span>: <span class="st">'Stroke Risk Guidelines'</span>,</span>
<span id="cb19-160"><a href="#cb19-160" aria-hidden="true" tabindex="-1"></a>                <span class="st">'urgency'</span>: <span class="st">'high'</span></span>
<span id="cb19-161"><a href="#cb19-161" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb19-162"><a href="#cb19-162" aria-hidden="true" tabindex="-1"></a>            <span class="st">'FTO'</span>: {</span>
<span id="cb19-163"><a href="#cb19-163" aria-hidden="true" tabindex="-1"></a>                <span class="st">'title'</span>: <span class="st">'FTO Risk Variant Detected'</span>,</span>
<span id="cb19-164"><a href="#cb19-164" aria-hidden="true" tabindex="-1"></a>                <span class="st">'detail'</span>: <span class="st">'Associated with obesity and insulin resistance'</span>,</span>
<span id="cb19-165"><a href="#cb19-165" aria-hidden="true" tabindex="-1"></a>                <span class="st">'source'</span>: <span class="st">'Metabolic Guidelines'</span>,</span>
<span id="cb19-166"><a href="#cb19-166" aria-hidden="true" tabindex="-1"></a>                <span class="st">'urgency'</span>: <span class="st">'medium'</span></span>
<span id="cb19-167"><a href="#cb19-167" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb19-168"><a href="#cb19-168" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb19-169"><a href="#cb19-169" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb19-170"><a href="#cb19-170" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Add recommendations for each significant gene</span></span>
<span id="cb19-171"><a href="#cb19-171" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> gene <span class="kw">in</span> genes:</span>
<span id="cb19-172"><a href="#cb19-172" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> gene <span class="kw">in</span> gene_recommendations:</span>
<span id="cb19-173"><a href="#cb19-173" aria-hidden="true" tabindex="-1"></a>                recommendations.append(gene_recommendations[gene])</span>
<span id="cb19-174"><a href="#cb19-174" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="ss">f"Added recommendation for gene: </span><span class="sc">{</span>gene<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb19-175"><a href="#cb19-175" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb19-176"><a href="#cb19-176" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Check for risk categories in all variants</span></span>
<span id="cb19-177"><a href="#cb19-177" aria-hidden="true" tabindex="-1"></a>        all_risks <span class="op">=</span> <span class="st">' '</span>.join(pathogenic_variants[<span class="st">'associated_risks'</span>].dropna())</span>
<span id="cb19-178"><a href="#cb19-178" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb19-179"><a href="#cb19-179" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Add general variant recommendations if no specific ones</span></span>
<span id="cb19-180"><a href="#cb19-180" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="kw">not</span> recommendations <span class="kw">and</span> <span class="bu">len</span>(pathogenic_variants) <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb19-181"><a href="#cb19-181" aria-hidden="true" tabindex="-1"></a>            unique_genes <span class="op">=</span> <span class="st">', '</span>.join(genes)</span>
<span id="cb19-182"><a href="#cb19-182" aria-hidden="true" tabindex="-1"></a>            recommendations.append({</span>
<span id="cb19-183"><a href="#cb19-183" aria-hidden="true" tabindex="-1"></a>                <span class="st">'title'</span>: <span class="st">'Genetic Variants Detected'</span>,</span>
<span id="cb19-184"><a href="#cb19-184" aria-hidden="true" tabindex="-1"></a>                <span class="st">'detail'</span>: <span class="ss">f'Clinically significant variants found in genes: </span><span class="sc">{</span>unique_genes<span class="sc">}</span><span class="ss">'</span>,</span>
<span id="cb19-185"><a href="#cb19-185" aria-hidden="true" tabindex="-1"></a>                <span class="st">'source'</span>: <span class="st">'Genetic Analysis'</span>,</span>
<span id="cb19-186"><a href="#cb19-186" aria-hidden="true" tabindex="-1"></a>                <span class="st">'urgency'</span>: <span class="st">'medium'</span></span>
<span id="cb19-187"><a href="#cb19-187" aria-hidden="true" tabindex="-1"></a>            })</span>
<span id="cb19-188"><a href="#cb19-188" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f"Added general recommendation for genes: </span><span class="sc">{</span>unique_genes<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb19-189"><a href="#cb19-189" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb19-190"><a href="#cb19-190" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"No pathogenic variants found, no recommendations generated"</span>)</span>
<span id="cb19-191"><a href="#cb19-191" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-192"><a href="#cb19-192" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Convert recommendations to CDS cards</span></span>
<span id="cb19-193"><a href="#cb19-193" aria-hidden="true" tabindex="-1"></a>    cards <span class="op">=</span> []</span>
<span id="cb19-194"><a href="#cb19-194" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-195"><a href="#cb19-195" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> rec <span class="kw">in</span> recommendations:</span>
<span id="cb19-196"><a href="#cb19-196" aria-hidden="true" tabindex="-1"></a>        cards.append({</span>
<span id="cb19-197"><a href="#cb19-197" aria-hidden="true" tabindex="-1"></a>            <span class="st">'summary'</span>: <span class="ss">f"</span><span class="sc">{</span>rec[<span class="st">'title'</span>]<span class="sc">}</span><span class="ss">"</span>,</span>
<span id="cb19-198"><a href="#cb19-198" aria-hidden="true" tabindex="-1"></a>            <span class="st">'indicator'</span>: <span class="st">'warning'</span> <span class="cf">if</span> rec[<span class="st">'urgency'</span>] <span class="op">==</span> <span class="st">'high'</span> <span class="cf">else</span> <span class="st">'info'</span>,</span>
<span id="cb19-199"><a href="#cb19-199" aria-hidden="true" tabindex="-1"></a>            <span class="st">'detail'</span>: rec[<span class="st">'detail'</span>],</span>
<span id="cb19-200"><a href="#cb19-200" aria-hidden="true" tabindex="-1"></a>            <span class="st">'source'</span>: {</span>
<span id="cb19-201"><a href="#cb19-201" aria-hidden="true" tabindex="-1"></a>                <span class="st">'label'</span>: rec[<span class="st">'source'</span>]</span>
<span id="cb19-202"><a href="#cb19-202" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb19-203"><a href="#cb19-203" aria-hidden="true" tabindex="-1"></a>        })</span>
<span id="cb19-204"><a href="#cb19-204" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-205"><a href="#cb19-205" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no recommendations were found, provide a default card</span></span>
<span id="cb19-206"><a href="#cb19-206" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> cards:</span>
<span id="cb19-207"><a href="#cb19-207" aria-hidden="true" tabindex="-1"></a>        cards <span class="op">=</span> [{</span>
<span id="cb19-208"><a href="#cb19-208" aria-hidden="true" tabindex="-1"></a>            <span class="st">'summary'</span>: <span class="ss">f'No Significant Genetic Findings - </span><span class="sc">{</span>test_patient_id<span class="sc">}</span><span class="ss">'</span>,</span>
<span id="cb19-209"><a href="#cb19-209" aria-hidden="true" tabindex="-1"></a>            <span class="st">'indicator'</span>: <span class="st">'info'</span>,</span>
<span id="cb19-210"><a href="#cb19-210" aria-hidden="true" tabindex="-1"></a>            <span class="st">'detail'</span>: <span class="st">'No pathogenic or likely pathogenic variants detected in this patient</span><span class="ch">\'</span><span class="st">s genetic analysis.'</span>,</span>
<span id="cb19-211"><a href="#cb19-211" aria-hidden="true" tabindex="-1"></a>            <span class="st">'source'</span>: {</span>
<span id="cb19-212"><a href="#cb19-212" aria-hidden="true" tabindex="-1"></a>                <span class="st">'label'</span>: <span class="st">'Genomics Service'</span></span>
<span id="cb19-213"><a href="#cb19-213" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb19-214"><a href="#cb19-214" aria-hidden="true" tabindex="-1"></a>        }]</span>
<span id="cb19-215"><a href="#cb19-215" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-216"><a href="#cb19-216" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Return the CDS Hooks response</span></span>
<span id="cb19-217"><a href="#cb19-217" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> jsonify({<span class="st">'cards'</span>: cards})</span>
<span id="cb19-218"><a href="#cb19-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-219"><a href="#cb19-219" aria-hidden="true" tabindex="-1"></a><span class="co"># Start Flask in a background thread</span></span>
<span id="cb19-220"><a href="#cb19-220" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> run_flask_in_thread():</span>
<span id="cb19-221"><a href="#cb19-221" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> werkzeug.serving <span class="im">import</span> run_simple</span>
<span id="cb19-222"><a href="#cb19-222" aria-hidden="true" tabindex="-1"></a>    run_simple(<span class="st">'localhost'</span>, flask_port, app, use_reloader<span class="op">=</span><span class="va">False</span>, use_debugger<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb19-223"><a href="#cb19-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-224"><a href="#cb19-224" aria-hidden="true" tabindex="-1"></a><span class="co"># Start the Flask server</span></span>
<span id="cb19-225"><a href="#cb19-225" aria-hidden="true" tabindex="-1"></a>flask_server_thread <span class="op">=</span> threading.Thread(target<span class="op">=</span>run_flask_in_thread)</span>
<span id="cb19-226"><a href="#cb19-226" aria-hidden="true" tabindex="-1"></a>flask_server_thread.daemon <span class="op">=</span> <span class="va">True</span>  <span class="co"># Thread will exit when the notebook exits</span></span>
<span id="cb19-227"><a href="#cb19-227" aria-hidden="true" tabindex="-1"></a>flask_server_thread.start()</span>
<span id="cb19-228"><a href="#cb19-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-229"><a href="#cb19-229" aria-hidden="true" tabindex="-1"></a><span class="co"># Give the server a moment to start</span></span>
<span id="cb19-230"><a href="#cb19-230" aria-hidden="true" tabindex="-1"></a>time.sleep(<span class="dv">1</span>)</span>
<span id="cb19-231"><a href="#cb19-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-232"><a href="#cb19-232" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"CDS Hooks service running at http://localhost:</span><span class="sc">{</span>flask_port<span class="sc">}</span><span class="ss">/cds-services"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://localhost:3299
INFO:werkzeug:Press CTRL+C to quit</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>CDS Hooks service running at http://localhost:3299/cds-services</code></pre>
</div>
</div>
</section>
<section id="testing-our-cds-hooks-service" class="level3">
<h3 class="anchored" data-anchor-id="testing-our-cds-hooks-service">Testing Our CDS Hooks Service</h3>
<p>Now that our CDS Hooks service is running, let’s test it by simulating how an EHR would interact with it. When a clinician opens a patient’s record in an EHR system, the EHR sends a request to all registered CDS services, allowing them to provide relevant recommendations.</p>
<div id="29e96b62" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's directly query for genetic reports and work with a patient that definitely has genetic data</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Querying for genetic diagnostic reports..."</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>genetic_reports_dict <span class="op">=</span> search.steal_bundles_to_dataframe(</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    resource_type<span class="op">=</span><span class="st">"DiagnosticReport"</span>,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    request_params<span class="op">=</span>{</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"code"</span>: <span class="st">"http://loinc.org|55232-3"</span>,  <span class="co"># Genetic analysis summary panel</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"_count"</span>: <span class="dv">5</span>,</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"_include"</span>: <span class="st">"DiagnosticReport:result"</span>,  <span class="co"># Include the Observation resources</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>    num_pages<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>    fhir_paths<span class="op">=</span>[</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>        (<span class="st">"id"</span>, <span class="st">"id"</span>),</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>        (<span class="st">"subject"</span>, <span class="st">"subject.reference"</span>),</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>        (<span class="st">"status"</span>, <span class="st">"status"</span>),</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>        (<span class="st">"issued"</span>, <span class="st">"issued"</span>),</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Display diagnostic reports</span></span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>genetic_reports_df <span class="op">=</span> genetic_reports_dict[<span class="st">"DiagnosticReport"</span>]</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Found </span><span class="sc">{</span><span class="bu">len</span>(genetic_reports_df)<span class="sc">}</span><span class="ss"> genetic reports"</span>)</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>display(genetic_reports_df.head())</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Display observations count if available</span></span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">"Observation"</span> <span class="kw">in</span> genetic_reports_dict:</span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a>    observations_df <span class="op">=</span> genetic_reports_dict[<span class="st">"Observation"</span>]</span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Found </span><span class="sc">{</span><span class="bu">len</span>(observations_df)<span class="sc">}</span><span class="ss"> genetic observations"</span>)</span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Select a patient of interest, as if a clinician selected a patient from the EHR</span></span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a>test_patient_id <span class="op">=</span> genetic_reports_df.loc[genetic_reports_df[<span class="st">"subject"</span>] <span class="op">==</span> <span class="st">"Patient/df860bc2-1943-237f-7445-ed960a1ef069"</span>, <span class="st">"subject"</span>].values[<span class="dv">0</span>]</span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Selected patient with genetic data: </span><span class="sc">{</span>test_patient_id<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Testing with patient ID: </span><span class="sc">{</span>test_patient_id<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Querying for genetic diagnostic reports...
http://localhost:8080/fhir/DiagnosticReport?_count=5&amp;_include=DiagnosticReport:result&amp;code=http://loinc.org|55232-3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Query &amp; Build DF (DiagnosticReport):   0%|          | 0/1 [00:00&lt;?, ?it/s]Query &amp; Build DF (DiagnosticReport): 100%|██████████| 1/1 [00:00&lt;00:00, 46.06it/s]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Found 5 genetic reports</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">id</th>
<th data-quarto-table-cell-role="th">subject</th>
<th data-quarto-table-cell-role="th">status</th>
<th data-quarto-table-cell-role="th">issued</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>53db09ba-0414-61e1-78d2-02b20a7092b8</td>
<td>Patient/837e80f6-a7a5-77f8-36aa-c7b8ff002c4b</td>
<td>final</td>
<td>2001-02-25T10:39:54.395-05:00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>c012aa9a-fe79-40bc-00ae-162c9906c6f8</td>
<td>Patient/39533e4a-f6f2-a144-ab37-6500460250dc</td>
<td>final</td>
<td>2017-02-21T03:33:40.769-05:00</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>27e51912-a805-26dc-b9c8-d88f131969a6</td>
<td>Patient/68c3ae0b-e298-62b7-5d3a-7936fd998fe0</td>
<td>final</td>
<td>1989-10-03T04:33:40.769-04:00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>a31d8d94-6acd-a5f0-618a-f1b38746f9e4</td>
<td>Patient/df860bc2-1943-237f-7445-ed960a1ef069</td>
<td>final</td>
<td>1992-02-10T08:04:51.390-05:00</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>1e5e18ae-bad4-3a45-b4e3-d2c7b7da66d9</td>
<td>Patient/7d9ba758-f0b8-3fc3-befa-f0e8e8fb6935</td>
<td>final</td>
<td>2007-11-23T16:23:59.928-05:00</td>
</tr>
</tbody>
</table>

</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Found 159 genetic observations
Selected patient with genetic data: Patient/df860bc2-1943-237f-7445-ed960a1ef069
Testing with patient ID: Patient/df860bc2-1943-237f-7445-ed960a1ef069</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Testing Locally vs.&nbsp;Production
</div>
</div>
<div class="callout-body-container callout-body">
<p>In production, EHR systems would send requests directly to your CDS Hooks service. Since we’re developing locally, we need to manually mock these requests to test our service without connecting to a real EHR system.</p>
</div>
</div>
<div id="5d5bdb17" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a request with the selected patient</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>mock_request <span class="op">=</span> {</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"hook"</span>: <span class="st">"patient-view"</span>,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"hookInstance"</span>: <span class="st">"d1577c69-dfbe-44ad-ba6d-3e05e953b2ea"</span>,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"context"</span>: {</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"patientId"</span>: test_patient_id</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Send the request to our CDS service</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Sending request to CDS service..."</span>)</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>service_response <span class="op">=</span> requests.post(</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>    <span class="ss">f"http://localhost:</span><span class="sc">{</span>flask_port<span class="sc">}</span><span class="ss">/cds-services/genomics-advisor"</span>, </span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>    json<span class="op">=</span>mock_request</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the JSON response from our service</span></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">CDS Service Response:"</span>)</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(json.dumps(service_response.json(), indent<span class="op">=</span><span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sending request to CDS service...

CDS Service Response:
{
  "cards": [
    {
      "detail": "Consider cardiovascular risk assessment and lipid-lowering therapy",
      "indicator": "info",
      "source": {
        "label": "Cardiovascular Guidelines"
      },
      "summary": "PON1 Risk Variant Detected"
    },
    {
      "detail": "Associated with obesity and insulin resistance",
      "indicator": "info",
      "source": {
        "label": "Metabolic Guidelines"
      },
      "summary": "FTO Risk Variant Detected"
    },
    {
      "detail": "Associated with inflammatory processes and stroke risk",
      "indicator": "warning",
      "source": {
        "label": "Stroke Risk Guidelines"
      },
      "summary": "CCL2 Pathogenic Variant Detected"
    }
  ]
}</code></pre>
</div>
</div>
<p>The response contains CDS cards with our recommendations. These cards would typically be displayed directly in the EHR interface to help clinicians make informed decisions. Let’s create a visual representation of how these cards might appear in an EHR system:</p>
<div id="df05b0c1" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the CDS cards from the response</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>cards <span class="op">=</span> service_response.json().get(<span class="st">'cards'</span>, [])</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a simple HTML visualization to show how these cards would appear in an EHR</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>html_output <span class="op">=</span> <span class="ss">f"""</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="ss">&lt;div style="font-family: Arial, sans-serif; border: 1px solid #ccc; border-radius: 5px; padding: 15px; max-width: 700px; margin: 0 auto;"&gt;</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="ss">    &lt;div style="border-bottom: 1px solid #ddd; padding-bottom: 10px; margin-bottom: 15px;"&gt;</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="ss">        &lt;h3 style="margin-top: 0;"&gt;EHR Patient View: </span><span class="sc">{</span>test_patient_id<span class="sc">}</span><span class="ss">&lt;/h3&gt;</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a><span class="ss">        &lt;div style="color: #666; font-size: 0.9em;"&gt;Genomic Decision Support&lt;/div&gt;</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a><span class="ss">    &lt;/div&gt;</span></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a><span class="ss">    &lt;div style="background-color: #f8f9fa; border: 1px solid #e9ecef; padding: 10px; margin-bottom: 15px; border-radius: 3px;"&gt;</span></span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a><span class="ss">        &lt;p style="margin: 0; font-size: 0.9em; color: #6c757d;"&gt;</span></span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a><span class="ss">            &lt;strong&gt;Demo Note:&lt;/strong&gt; This is a simulation of how CDS Hooks would appear in an EHR. </span></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a><span class="ss">            The buttons shown below are not functional in this demo. In a real implementation, </span></span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a><span class="ss">            clinicians would be able to click these buttons to take recommended actions.</span></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a><span class="ss">        &lt;/p&gt;</span></span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a><span class="ss">    &lt;/div&gt;</span></span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a><span class="ss">"""</span></span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Add each card to the visualization</span></span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> cards:</span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>    html_output <span class="op">+=</span> <span class="st">"&lt;p&gt;No clinical decision support recommendations available for this patient.&lt;/p&gt;"</span></span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> card <span class="kw">in</span> cards:</span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Set card color based on importance</span></span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a>        border_color <span class="op">=</span> <span class="st">"#007bff"</span>  <span class="co"># Default blue</span></span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> card.get(<span class="st">'indicator'</span>) <span class="op">==</span> <span class="st">"warning"</span>:</span>
<span id="cb30-29"><a href="#cb30-29" aria-hidden="true" tabindex="-1"></a>            border_color <span class="op">=</span> <span class="st">"#ff9800"</span>  <span class="co"># Orange for warnings</span></span>
<span id="cb30-30"><a href="#cb30-30" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> card.get(<span class="st">'indicator'</span>) <span class="op">==</span> <span class="st">"critical"</span>:</span>
<span id="cb30-31"><a href="#cb30-31" aria-hidden="true" tabindex="-1"></a>            border_color <span class="op">=</span> <span class="st">"#dc3545"</span>  <span class="co"># Red for critical alerts</span></span>
<span id="cb30-32"><a href="#cb30-32" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb30-33"><a href="#cb30-33" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Create the card HTML</span></span>
<span id="cb30-34"><a href="#cb30-34" aria-hidden="true" tabindex="-1"></a>        html_output <span class="op">+=</span> <span class="ss">f"""</span></span>
<span id="cb30-35"><a href="#cb30-35" aria-hidden="true" tabindex="-1"></a><span class="ss">        &lt;div style="border-left: 4px solid </span><span class="sc">{</span>border_color<span class="sc">}</span><span class="ss">; padding: 10px 15px; margin-bottom: 10px; background-color: #f8f9fa;"&gt;</span></span>
<span id="cb30-36"><a href="#cb30-36" aria-hidden="true" tabindex="-1"></a><span class="ss">            &lt;h4 style="margin-top: 0;"&gt;</span><span class="sc">{</span>card<span class="sc">.</span>get(<span class="st">'summary'</span>, <span class="st">''</span>)<span class="sc">}</span><span class="ss">&lt;/h4&gt;</span></span>
<span id="cb30-37"><a href="#cb30-37" aria-hidden="true" tabindex="-1"></a><span class="ss">            &lt;p&gt;</span><span class="sc">{</span>card<span class="sc">.</span>get(<span class="st">'detail'</span>, <span class="st">''</span>)<span class="sc">}</span><span class="ss">&lt;/p&gt;</span></span>
<span id="cb30-38"><a href="#cb30-38" aria-hidden="true" tabindex="-1"></a><span class="ss">        """</span></span>
<span id="cb30-39"><a href="#cb30-39" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb30-40"><a href="#cb30-40" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Add suggestion buttons</span></span>
<span id="cb30-41"><a href="#cb30-41" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="st">'suggestions'</span> <span class="kw">in</span> card:</span>
<span id="cb30-42"><a href="#cb30-42" aria-hidden="true" tabindex="-1"></a>            html_output <span class="op">+=</span> <span class="st">'&lt;div&gt;'</span></span>
<span id="cb30-43"><a href="#cb30-43" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> suggestion <span class="kw">in</span> card.get(<span class="st">'suggestions'</span>, []):</span>
<span id="cb30-44"><a href="#cb30-44" aria-hidden="true" tabindex="-1"></a>                html_output <span class="op">+=</span> <span class="ss">f'&lt;span style="display: inline-block; background: #e9ecef; border: 1px solid #ced4da; padding: 5px 10px; border-radius: 3px; margin-top: 8px; margin-right: 5px; cursor: pointer;"&gt;</span><span class="sc">{</span>suggestion<span class="sc">.</span>get(<span class="st">"label"</span>, <span class="st">""</span>)<span class="sc">}</span><span class="ss">&lt;/span&gt;'</span></span>
<span id="cb30-45"><a href="#cb30-45" aria-hidden="true" tabindex="-1"></a>            html_output <span class="op">+=</span> <span class="st">'&lt;/div&gt;'</span></span>
<span id="cb30-46"><a href="#cb30-46" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb30-47"><a href="#cb30-47" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Add source information</span></span>
<span id="cb30-48"><a href="#cb30-48" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="st">'source'</span> <span class="kw">in</span> card:</span>
<span id="cb30-49"><a href="#cb30-49" aria-hidden="true" tabindex="-1"></a>            html_output <span class="op">+=</span> <span class="ss">f'&lt;div style="color: #6c757d; font-size: 0.85em; margin-top: 5px;"&gt;Source: </span><span class="sc">{</span>card<span class="sc">.</span>get(<span class="st">"source"</span>, {})<span class="sc">.</span>get(<span class="st">"label"</span>, <span class="st">""</span>)<span class="sc">}</span><span class="ss">&lt;/div&gt;'</span></span>
<span id="cb30-50"><a href="#cb30-50" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb30-51"><a href="#cb30-51" aria-hidden="true" tabindex="-1"></a>        html_output <span class="op">+=</span> <span class="st">'&lt;/div&gt;'</span></span>
<span id="cb30-52"><a href="#cb30-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-53"><a href="#cb30-53" aria-hidden="true" tabindex="-1"></a>html_output <span class="op">+=</span> <span class="st">"&lt;/div&gt;"</span></span>
<span id="cb30-54"><a href="#cb30-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-55"><a href="#cb30-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the HTML</span></span>
<span id="cb30-56"><a href="#cb30-56" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">Visualization of CDS Cards in EHR Interface:"</span>)</span>
<span id="cb30-57"><a href="#cb30-57" aria-hidden="true" tabindex="-1"></a>display(HTML(html_output))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Visualization of CDS Cards in EHR Interface:</code></pre>
</div>
<div class="cell-output cell-output-display">

<div style="font-family: Arial, sans-serif; border: 1px solid #ccc; border-radius: 5px; padding: 15px; max-width: 700px; margin: 0 auto;">
    <div style="border-bottom: 1px solid #ddd; padding-bottom: 10px; margin-bottom: 15px;">
        <h3 style="margin-top: 0;" class="anchored">EHR Patient View: Patient/df860bc2-1943-237f-7445-ed960a1ef069</h3>
        <div style="color: #666; font-size: 0.9em;">Genomic Decision Support</div>
    </div>
    
    <div style="background-color: #f8f9fa; border: 1px solid #e9ecef; padding: 10px; margin-bottom: 15px; border-radius: 3px;">
        <p style="margin: 0; font-size: 0.9em; color: #6c757d;">
            <strong>Demo Note:</strong> This is a simulation of how CDS Hooks would appear in an EHR. 
            The buttons shown below are not functional in this demo. In a real implementation, 
            clinicians would be able to click these buttons to take recommended actions.
        </p>
    </div>

        <div style="border-left: 4px solid #007bff; padding: 10px 15px; margin-bottom: 10px; background-color: #f8f9fa;">
            <h4 style="margin-top: 0;" class="anchored">PON1 Risk Variant Detected</h4>
            <p>Consider cardiovascular risk assessment and lipid-lowering therapy</p>
        <div style="color: #6c757d; font-size: 0.85em; margin-top: 5px;">Source: Cardiovascular Guidelines</div></div>
        <div style="border-left: 4px solid #007bff; padding: 10px 15px; margin-bottom: 10px; background-color: #f8f9fa;">
            <h4 style="margin-top: 0;" class="anchored">FTO Risk Variant Detected</h4>
            <p>Associated with obesity and insulin resistance</p>
        <div style="color: #6c757d; font-size: 0.85em; margin-top: 5px;">Source: Metabolic Guidelines</div></div>
        <div style="border-left: 4px solid #ff9800; padding: 10px 15px; margin-bottom: 10px; background-color: #f8f9fa;">
            <h4 style="margin-top: 0;" class="anchored">CCL2 Pathogenic Variant Detected</h4>
            <p>Associated with inflammatory processes and stroke risk</p>
        <div style="color: #6c757d; font-size: 0.85em; margin-top: 5px;">Source: Stroke Risk Guidelines</div></div></div>
</div>
</div>
</section>
</section>
<section id="summary-and-next-steps" class="level2">
<h2 class="anchored" data-anchor-id="summary-and-next-steps">Summary and Next Steps</h2>
<p>Congratulations on completing this tutorial on genomic clinical decision support! We’ve explored how to build a CDS service that can identify clinically relevant genomic variants and deliver actionable recommendations to clinicians at the point of care.</p>
<p>In this tutorial, we’ve covered:</p>
<ol type="1">
<li><strong>Basic genomic data retrieval</strong> from FHIR servers</li>
<li><strong>Variant analysis</strong> and extraction of clinically relevant information</li>
<li><strong>Implementation of a CDS Hooks service</strong> to deliver genomic recommendations</li>
<li><strong>Integration simulation</strong> showing how the service would appear in an EHR</li>
</ol>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Critical Implementation Considerations for Production Deployment
</div>
</div>
<div class="callout-body-container callout-body">
<p>Moving a prototype CDS service to production involves significant security, compliance, and safety concerns. Improper implementation could pose risks to patient care and data security.</p>
<p>To extend this work in a production environment, you would need to:</p>
<ol type="1">
<li><strong>Host the Service</strong>:
<ul>
<li>Deploy on a secure, HIPAA-compliant infrastructure with HTTPS and appropriate security measures</li>
</ul></li>
<li><strong>Implement Authentication</strong>:
<ul>
<li>Use industry-standard authentication protocols supported by your EHR</li>
</ul></li>
<li><strong>Update Clinical Knowledge</strong>:
<ul>
<li>Store clinical recommendations in a database rather than hardcoded values</li>
<li>Establish a regular review cycle with clinical experts</li>
</ul></li>
<li><strong>Add Robust Error Handling</strong>:
<ul>
<li>Implement comprehensive exception handling, logging, and monitoring</li>
<li>Develop a comprehensive test suite</li>
</ul></li>
<li><strong>Test with Users</strong>:
<ul>
<li>Conduct usability testing with actual clinicians</li>
<li>Start with a limited pilot before full deployment</li>
<li>Implement feedback mechanisms</li>
</ul></li>
</ol>
<p>Remember that clinical decision support systems directly impact patient care—making robust implementation both an ethical and legal obligation.</p>
</div>
</div>
</section>
<section id="additional-resources" class="level2">
<h2 class="anchored" data-anchor-id="additional-resources">Additional Resources</h2>
<p>For more information, see:</p>
<ul>
<li><a href="http://hl7.org/fhir/uv/genomics-reporting/">FHIR Genomics Implementation Guide</a></li>
<li><a href="https://cds-hooks.org/">CDS Hooks Documentation</a></li>
<li><a href="https://loinc.org/genetic-variant/">LOINC Genetic Variant Terms</a></li>
<li><a href="../../modules/new-cds-hooks-intro">Learn more about CDS Hooks</a></li>
</ul>


</section>

</main> <!-- /main -->
<!-- scripts.html -->
<!-- any content in this file is inserted right before </body> -->

<!-- Append index.html to any links ending with '/'

Example:

<a href="https://example.org/workshops/bulk-data/">

will turn into:

<a href="https://example.org/workshops/bulk-data/index.html">

This ensures consistent behavior across different deployment platforms
and enables other scripts like role-based module highlighting.
-->
<script>
  document.addEventListener("DOMContentLoaded", function() {
    document.querySelectorAll("a[href]").forEach(link => {
      if( link.getAttribute("href").endsWith("/") ) {
        link.setAttribute("href", link.getAttribute("href") + "index.html");
      }
    })
  });
</script>

<!-- Fix a11y issues with side menu -->
<script>
const elements = document.querySelectorAll('.sidebar-menu-container a.sidebar-item-text[data-bs-toggle="collapse"], .sidebar-menu-container a.sidebar-item-toggle[data-bs-toggle="collapse"]')

elements.forEach(element => {
    element.setAttribute('role', 'group');
});
</script>

<!-- JSON Viewer
Intercept all instances of:

``` json
```

and replace it with a fancy JSON viewer. If the content is
invalid JSON then the default Quarto rendering will occur.
-->
<script type="text/javascript" src="https://unpkg.com/dompurify@3.0.1/dist/purify.min.js"></script> <!-- load JS Sanitizer -->
<script>
    function jsonViewer(json, collapsible=false) {
        let TEMPLATES = {
            item: '<div class="json_view__item"><div class="json_view__key">%KEY%</div><div class="json_view__value json_view__value--%TYPE%">%VALUE%</div></div>',
            itemCollapsible: '<label class="json_view__item json_view__item--collapsible"><input type="checkbox" class="json_view__toggle"/><div class="json_view__key">%KEY%</div><div class="json_view__value json_view__value--type-%TYPE%">%VALUE%</div>%CHILDREN%</label>',
            itemCollapsibleOpen: '<label class="json_view__item json_view__item--collapsible"><input type="checkbox" checked class="json_view__toggle"/><div class="json_view__key">%KEY%</div><div class="json_view__value json_view__value--type-%TYPE%">%VALUE%</div>%CHILDREN%</label>'
        };

        function createItem(key, value, type){
            let element = TEMPLATES.item.replace('%KEY%', key);

            if((key == 'div') && (type=='string')) { // Removes Resource.text.div value (looks bloated)
                element = element.replace('%VALUE%', '...');
            } else if(type == 'string') {
                element = element.replace('%VALUE%', '"' + value + '"');
            } else {
                element = element.replace('%VALUE%', value);
            }

            element = element.replace('%TYPE%', type);

            return element;
        }

        function createCollapsibleItem(key, value, type, children){
            let tpl = 'itemCollapsible';

            if(collapsible) {
                tpl = 'itemCollapsibleOpen';
            }

            let element = TEMPLATES[tpl].replace('%KEY%', key);

            element = element.replace('%VALUE%', type);
            element = element.replace('%TYPE%', type);
            element = element.replace('%CHILDREN%', children);

            return element;
        }

        function handleChildren(key, value, type) {
            let html = '';

            for(let item in value) { 
                let _key = item,
                    _val = value[item];

                html += handleItem(_key, _val);
            }

            return createCollapsibleItem(key, value, type, html);
        }

        function handleItem(key, value) {
            let type = typeof value;
 
            if( Array.isArray( value ) ) {
                return handleChildren(key, value, "array");
            }

            if(typeof value === 'object') {        
                return handleChildren(key, value, type);
            }

            return createItem(key, value, type);
        }

        function parseObject(obj) {
            _result = '<div class="json_view">';

            for(let item in obj) { 
                let key = item,
                    value = obj[item];

                _result += handleItem(key, value);
            }

            _result += '</div>';

            return _result;
        }
        
        try {
            return parseObject(JSON.parse(json));
        }
        catch(err) {
            console.log("Could not render JSON viewer: ", err);
            console.log("for string: ", json);
            return null;
        }

    };


    function parseHTML(string) {
        let sanitized = DOMPurify.sanitize(string);
        let parser = new DOMParser();
        return parser.parseFromString(sanitized, "text/html").body.firstChild;
    }

    let jsonCodeBlocks = document.querySelectorAll("code.sourceCode.json");
    jsonCodeBlocks.forEach((jsonCode) => {
        let json = jsonCode.innerText;
        let codeBlock = jsonCode.parentNode.parentNode;
        let uid = codeBlock.id;
        let documentLocation = codeBlock.parentNode;
        let viewerHTML = jsonViewer(json, true);

        let tabContainer = parseHTML(`
          <div>
            <ul class="nav nav-tabs" id="${uid}Tabs" role="tablist">
              <li class="nav-item" role="presentation">
                <button class="nav-link active" id="${uid}ViewerTab" data-bs-toggle="tab" data-bs-target="#${uid}Viewer" type="button" role="tab" aria-controls="${uid}Viewer" aria-selected="true">Viewer</button>
              </li>
              <li class="nav-item" role="presentation">
                <button class="nav-link" id="${uid}JsonTab" data-bs-toggle="tab" data-bs-target="#${uid}Json" type="button" role="tab" aria-controls="${uid}Json" aria-selected="false">JSON</button>
              </li>
            </ul>
            <div class="tab-content" id="${uid}TabContent" style="height:50vh;overflow:scroll;">
              <div class="tab-pane fade show active" id="${uid}Viewer" role="tabpanel" aria-labelledby="${uid}ViewerTab"></div>
              <div class="tab-pane fade" id="${uid}Json" role="tabpanel" aria-labelledby="${uid}JsonTab"></div>
            </div>
          </div>
        `)

        if( viewerHTML != null ) {
            documentLocation.replaceChild(tabContainer, codeBlock);

            tabContainer.querySelector(`div#${uid}TabContent div#${uid}Viewer`).appendChild( parseHTML(viewerHTML) );
            tabContainer.querySelector(`div#${uid}TabContent div#${uid}Json`).appendChild(codeBlock);
        }
    });
</script>

<!-- Role-based Module Highlighting

For any *.qmd file with roles defined in frontmatter,
running script/update-module-role-map will generate
modules/mappings.js. Then any markup with class .highlight-by-role, i.e:

:::{.highlight-by-role}
 - [mod1](/mod1)
 - [mod2](/mod2)
 - [mod3](/mod3)
:::

will highlight the links in the list assigned to the users roles.

The user can select a role by pressing a button with data-role=*
attribute. The possible roles are:

Possible Roles:
  investigator
  research-leader
  informaticist
  software-engineer
  clinician-scientist
-->
<script>
const all_modules = document.querySelectorAll(".highlight-by-role ul li");

/* highlights modules corresponding to role via role_module_map
 * @param: string - role name in lower kebab case
 */
function select_modules(string) {
    // remove all highlights
    all_modules.forEach((li_node) => {
        var mark_node = li_node.firstChild;
        var a_node = mark_node.firstChild;

        if( mark_node.nodeName == "MARK" ) {
            li_node.removeChild(mark_node);
            li_node.appendChild(a_node);
        }
    });

    // add highlights based on role_module_map
    role_module_map.forEach((mapping) => {
        if( mapping.role == string ) {
            mapping.modules.forEach((interest_module) => {
                document.querySelectorAll(`li [href*="${interest_module.slug}"]`).forEach((interest_a_node) => {

                    //console.log("selecting module ", interest_a_node);

                    var interest_li_node = interest_a_node.parentNode;
                    var mark_node = document.createElement("mark");
                    interest_li_node.appendChild(mark_node);
                    mark_node.appendChild(interest_a_node);

                    //console.log("appended ", interest_a_node, " to ", mark_node);
                });
            });
            return;
        }
    })
}


// clears previously selected roles, highlights the selected role, and sets "role" in local storage
function select_role(role) {
    store.setItem("role", role);

    document.querySelectorAll("[data-role]").forEach((a_node) => {
        a_node.classList.remove("btn", "btn-sm", "btn-secondary", "fw-bold", "rounded");
        a_node.classList.add("btn-role");
    });

    var selected_a_node = document.querySelector(`[data-role="${role}"]`);
    selected_a_node.classList.add("btn", "btn-sm", "btn-secondary", "fw-bold", "rounded");
    selected_a_node.classList.remove("btn-role");
}


// trigger role selection on clicking <a data-role=...>
document.querySelectorAll("a[data-role]").forEach((role_link) => {
    role_link.addEventListener("click", (e) => {
        var role = e.target.dataset.role;
        select_role(role);
        select_modules(role);
    });
});

// set default role and execute select_role and select_modules once to make sure feature is working
window.addEventListener("load", () => {
    if( store.getItem("role") ) {
        select_role(store.getItem("role"));
        select_modules(store.getItem("role"));
    }
    else {
        const DEFAULT_ROLE = "research-leader";

        select_role(DEFAULT_ROLE);
        select_modules(DEFAULT_ROLE);
    }
});
</script>

<!-- Navigation Override

Overwrite Quarto's Next/Previous buttons to point to modules
for the user's role
-->
<script>
// Below function is usable by CC BY-SA 3.0 from https://stackoverflow.com/a/6475125
String.prototype.toTitleCase = function() {
  let i, j, str, lowers, uppers;
  str = this.replace(/([^\W_]+[^\s-]*) */g, function(txt) {
    return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
  });

  // Certain minor words should be left lowercase unless 
  // they are the first or last words in the string
  lowers = ['A', 'An', 'The', 'And', 'But', 'Or', 'For', 'Nor', 'As', 'At', 
  'By', 'For', 'From', 'In', 'Into', 'Near', 'Of', 'On', 'Onto', 'To', 'With'];
  for (i = 0, j = lowers.length; i < j; i++)
    str = str.replace(new RegExp('\\s' + lowers[i] + '\\s', 'g'), 
      function(txt) {
        return txt.toLowerCase();
      });

  // Certain words such as initialisms or acronyms should be left uppercase
  uppers = ['ID', 'TV', 'FHIR'];
  for (i = 0, j = uppers.length; i < j; i++)
    str = str.replace(new RegExp('\\b' + uppers[i] + '\\b', 'g'), 
      uppers[i].toUpperCase());

  return str;
}

// Custom Hack to overwrite Quarto's built in navigators
document.addEventListener("DOMContentLoaded", function() {
    // Get current module and role
    const current_module_slug = window.location.toString().split('/').pop().split('.')[0];
    const role = store.getItem("role");

    // Get next and prev modules data
    let prev_href, prev_text;
    let next_href, next_text;

    role_module_map.forEach((mapping) => {
        if( mapping.role == role ) {
            for(let i=0; i<mapping.modules.length; i++) {
                if( mapping.modules[i].slug == current_module_slug ) {
                    function slug_to_url(slug) {
                        if( window.location.pathname.includes('modules') ) {
                            // replace "modules/*" in current URL with "modules/<slug>.html"
                            return window.location.toString().replace(/modules\/.*/, "modules/" + slug + ".html");
                        }
                        else {
                            // this case is irrelevant since prev/next buttons are only role-based in modules
                            return window.location.toString() + "/modules/" + slug + ".html";
                        }
                    }

                    prev_href = (i > 0) ? slug_to_url(mapping.modules[i-1].slug) : null;
                    prev_text = (i > 0) ? mapping.modules[i-1].text : null;

                    next_href = (i+1 < mapping.modules.length) ? slug_to_url(mapping.modules[i+1].slug) : null;
                    next_text = (i+1 < mapping.modules.length) ? mapping.modules[i+1].text : null;
                }
            }
        }
    });


    // Overwrite prev link
    if( prev_href ) {
        const prev_link_elmt = document.querySelector(".page-navigation .nav-page-previous .pagination-link");
        const prev_text_elmt = document.querySelector(".page-navigation .nav-page-previous .pagination-link .nav-page-text");
        if( prev_link_elmt ) {

            prev_link_elmt.href = prev_href
            //prev_link_elmt.classList.add("btn", "btn-outline-custom", "rounded");
            //prev_link_elmt.parentNode.classList.toggle("nav-page");
            prev_text_elmt.innerText = prev_text;
        }
    }

    // Overwrite next link
    if( next_href ) {
        const next_link_elmt = document.querySelector(".page-navigation .nav-page-next .pagination-link");
        const next_text_elmt = document.querySelector(".page-navigation .nav-page-next .pagination-link .nav-page-text");

        if( next_link_elmt ) {
            next_link_elmt.href = next_href;
            //next_link_elmt.classList.add("btn", "btn-outline-custom", "rounded");
            //next_link_elmt.parentNode.classList.toggle("nav-page");
            next_text_elmt.innerText = next_text;
        }
    }

    // if document has #key-points then move pagination to next after it
    key_points = document.querySelector("#key-points");
    if( key_points ) {
        let page_nav = document.querySelector("nav.page-navigation");
        key_points.insertAdjacentElement("afterend", page_nav);
    }

});


// Override style of Quarto page navigation (prev/next buttons)
document.addEventListener("DOMContentLoaded", function() {

    const page_navigators = document.querySelectorAll(".pagination-link");
    page_navigators.forEach((node) => {

        node.classList.add("btn", "btn-outline-custom", "rounded");
        node.parentNode.classList.remove("nav-page");
        node.parentNode.parentNode.style.gridRow = "initial";
    });
});

// Set external links to open in a new tab
document.addEventListener("DOMContentLoaded", function() {
    var all_links = document.querySelectorAll('a');
    for (var i = 0; i < all_links.length; i++){
        var a = all_links[i];
        if(a.hostname != location.hostname) {
                a.target = '_blank';
        }
    }
})

</script>

<!-- Lightbox 

Intercept all instances of:

:::{#id .mermaid-lightbox data-caption='my caption'}
```{mermaid}
mermaid markdown content...
```
:::

and overwrite it with a popup lightbox for easy viewing of diagram.
Powered by GLightbox: https://github.com/biati-digital/glightbox
-->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/glightbox@3.3.1/dist/css/glightbox.min.css">
<script src="https://cdn.jsdelivr.net/gh/mcstudios/glightbox@3.3.1/dist/js/glightbox.min.js"></script>
<script>
  document.addEventListener("DOMContentLoaded", function() {
    const mermaidElements = document.querySelectorAll('.mermaid-lightbox');

    mermaidElements.forEach(element => {
      const id = element.id;
      const caption = element.getAttribute('data-caption');

      const lightboxFigure = document.createElement('a');
      lightboxFigure.setAttribute('href', '#' + id);
      lightboxFigure.className = 'glightbox';
      lightboxFigure.setAttribute('data-description', caption);
      lightboxFigure.setAttribute('data-width', "90%");
      lightboxFigure.setAttribute('data-height', "90%");

      const clonedElement = element.cloneNode(true);
      lightboxFigure.appendChild(clonedElement);

      const figcaption = document.createElement('figcaption');
      figcaption.textContent = 'Click diagram to enlarge';
      lightboxFigure.appendChild(figcaption);

      element.parentNode.replaceChild(lightboxFigure, element);
    });

    // Initialize GLightbox
    const lightbox = GLightbox({
        selector: '.glightbox'
    });
  })
</script>

<!-- This script dynamically updates the copyright year in the footer -->
<script>
    document.addEventListener('DOMContentLoaded', function() {
    // Get the current year
    var currentYear = new Date().getFullYear();
    
    // Find the span element where we'll put the year
    var span = document.getElementById('current-year');
    
    if (span) {
        // If the span exists, update its content with the current year
        span.textContent = currentYear;
    }
    });
</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/nih-odss\.github\.io\/fhir-for-research\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../workshops/data-analysis-cds/01-fhir-analysis" class="pagination-link" aria-label="Part 1: FHIR Data Analysis">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Part 1: FHIR Data Analysis</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>© <span id="current-year"></span> The MITRE Corporation / <a href="https://github.com/NIH-ODSS/fhir-for-research/README.md">License information</a><br>Approved for Public Release / Case #24-3701<br><br>HL7® and FHIR® are the registered trademarks of <a href="https://hl7.org">Health Level Seven International (HL7)</a>.<br>Use of the HL7 and FHIR trademarks does not constitute an HL7 endorsement of this website.</p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/nih-odss/fhir-for-research/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>